### PR TITLE
[core] Modify quake.subveq such that it folds constant arguments.

### PR DESCRIPTION
**Pull request by schweitzpgi**
There was some technical debt with quake.subveq where constant operands were not folded into the operation like extract_ref and many others.

### CODE REVIEW

**Comment by schweitzpgi** [comment]
ping for review

File: include/cudaq/Optimizer/Dialect/Quake/Canonical.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace quake::canonical {
+
+inline mlir::Value createCast(mlir::PatternRewriter &rewriter,
+                              mlir::Location loc, mlir::Value inVal) {
+  auto i64Ty = rewriter.getI64Type();
+  assert(inVal.getType() != rewriter.getIndexType() &&
+         "use of index type is deprecated");
```
Comment (bmhowe23) [inline include/cudaq/Optimizer/Dialect/Quake/Canonical.h:23]: It looks like this code was moved, and it was "deprecated" before. Should we promote it to full-on "unsupported" now?
```suggestion
         "use of index type is unsupported");
```

**Review by bmhowe23 — APPROVED** [review]
LGTM 👍

**Review by annagrin — APPROVED** [review]
LGTM!

**Pull request merged**